api/tests/test_user_program_record.rs

```diff
@@ -76,3 +79,103 @@ async fn test_load_pbs_x() {
+        let prog_spec = ProgramSpecFactory::default()
+            .hardware_spec(&hw_spec)
+            .name("Prog1")
+            .insert(&conn);
+
+        // this should be deleted after the last record since it will no longer
+        // has any pbs.
+        let to_del = UserProgramRecordFactory::default()
+            .user(&user)
+            .program_spec(&prog_spec)
+            .cpu_cycles(100)
+            .instructions(100)
+            .registers_used(6)
+            .stacks_used(7)
+            .insert(&conn);
+
+        // This is also a trash record but it will be referenced so it will stay
+        let to_keep = UserProgramRecordFactory::default()
+            .user(&user)
+            .program_spec(&prog_spec)
+            .cpu_cycles(100)
+            .instructions(100)
+            .registers_used(6)
+            .stacks_used(7)
+            .insert(&conn);
+
+        UserProgramFactory::default()
+            .user(&user)
+            .program_spec(&prog_spec)
+            .file_name("existing.gdlk")
+            .source_code("READ RX0")
+            .record(Some(&to_keep))
+            .insert(&conn);
+
+        let cycle_pb1 = UserProgramRecordFactory::default()
+            .user(&user)
+            .program_spec(&prog_spec)
+            // These 2 are PBs
+            .cpu_cycles(1)
+            .instructions(2)
+            // These 2 are NOT PBs
+            .registers_used(10)
+            .stacks_used(10)
+            .insert(&conn);
+
+        let cycle_pb2 = UserProgramRecordFactory::default()
+            .user(&user)
+            .program_spec(&prog_spec)
+            // This is a PB thats the same as the previous one so it will not be
+            // deleted
+            .cpu_cycles(1)
+            // These 3 are NOT PBs
+            .instructions(100)
+            .registers_used(10)
+            .stacks_used(10)
+            .insert(&conn);
+        UserProgramRecordFactory::default()
+            .user(&user)
+            .program_spec(&prog_spec)
+            // These 2 are NOT PBs
+            .cpu_cycles(10)
+            .instructions(10)
+            // These 2 are PBs
+            .registers_used(3)
+            .stacks_used(4)
+            .insert(&conn);
+
+        // all records are either pbs or referenced by a user program
+        assert_eq!(
+            user_program_records::table
+                .select(dsl::count_star())
+                .get_result::<i64>(conn)
+                .unwrap(),
+            4
+        );
+
+        let remaining_ids: HashSet<Uuid> = user_program_records::table
+            .select(user_program_records::columns::id)
+            .get_results::<Uuid>(conn)
+            .unwrap()
+            .into_iter()
+            .collect();
+
+        assert!(!remaining_ids.contains(&to_del.id));
+        assert!(remaining_ids.contains(&to_keep.id));
+        assert!(remaining_ids.contains(&cycle_pb1.id));
+        assert!(remaining_ids.contains(&cycle_pb2.id));
```

Comment:
I think it would be cleaner to just do:

```rust
assert_eq!(
    remaining_ids,
    hashset! { to_del.id, to_keep.id, cycle_pb1.id, cycle_pb2.id }
);
```

(`hashset!` comes from the `maplit` crate)